README.md
# pi-power-relay
power cycle a device from a Raspberry Pi when network connectivity lost

## Description
If host(s) on an outside network are not reachable via ping, then reset a 
device, possibly a cable-modem, that is known to be flakey, by power cycling
it via a power relay.

This is intended to be run on a Raspberry Pi, using an 'IoT Power Relay
version 2' from Digital Loggers, to use GPIO to power-cycle a device.
The power relay is plugged into a UPS, while the Raspberry Pi is plugged
into the 'always ON' power port and the cable-modem is plugged into the
'normally ON' power port.  The program, when detecting a loss of network
connectivity, will drop power on the 'normally ON' port, and bring it 
back after a short period.

The power relay product used by this program can be seen at:
    https://dlidirect.com/products/iot-power-relay

This is typically run from the crontab every couple of minutes.

If run on something other than a Raspberry Pi, it will assume it is a
development or test box and will NOT attempt to do the actual shutdown,
but will still do everything else.

## Example usages
    % pi-power-relay --hosts 208.67.222.222,8.8.8.8 

This will ping the hosts 208.67.222.222 and 8.8.8.8.  If either of them
responds back correctly, then the program will exit with no action.

    % pi-power-relay --maint 01:20-01:40

This will prevent a reset of the device if the network goes away from 1:20am to 1:40am
during a known network maintenance period.
    
    % pi-power-relay --device-name 'cable-modem' --logfile /var/log/power-relay

This will log any resets to the log file /var/log/power-relay and will use the
device name of 'cable-modem' in the log entry.

    % pi-power-relay --delay-exit 180

The reason for this option is that if you do not use the --quiet option,
then an informational message is printed out before
the program exits.  But if you are relying on output from the crontab
to then be sent to you on some other machine via e-mail, it will fail
if you do not have a full blown mailer set up and the DNS lookup for
the destination will fail because the device (cable-modem?) has not
recovered yet.  So, this delay will give the device time to recover,
before you spit out the message and then exit.  The default is no delay.

There is a help option.  For eg:

    % pi-power-relay --help

    usage: pi-power-relay [options]*
        [-d|--debug]               debugging output
        [-e|--delay-exit num]      delay time before output and exit (0 secs)
        [-f|--force-reset]         reset now and quit, despite state or lock
        [-h|--help]                print this help info
        [-l|--logfile string]      log filename (none by default)
        [-m|--maint HH:MM-HH:MM]*  Maintenance time to NOT reset (none by default)
        [-p|--pin num ]            GPIO pin number (25)
        [-q|--quiet]               supress informational messages
        [-r|--reset-time num]      wait time between GPIO state change (15 secs)
        [-t|--tries num]           max number of ping attempts per host (3)
        [-w|--wait-time num]       reset wait time after previous reset (600 secs)
        [-x|--ping-timeout num]    wait time for ping to time out (2 secs)
        [-D|--device-name string]  name of thing being reset for log (device)
        [-H|--hosts string(s)]     comma-delimited hosts to ping (8.8.4.4,8.8.8.8)
        [-L|--lockfile string]     lock-file (/tmp/pi-power-relay--reset-time)
        [-V|--version]             print version of this program (2.0.2) 


## Python 2
The code will work with both Python2 and Python3, but the packaging and
distribution of this code is strongly tied to Python3.  If you are using
an ancient Python2 system, see the file README.install-python2 included
in the downloadable source code.


RELEASE_NOTES
1.0     June 9, 2018
        - initial version

1.1     June 22, 2018
        - fixed and improved some diagnostic messages
        - added -x|--ping-timeout option to change ping timeout from
          default of 2 seconds
        - added a sdist target in the Makefile

1.2     Oct 28, 2018
         - added a -m|--maint option to specify a known maintenance time
           that your outside network connection will not be available.
           Can be given multiple times.
           During these time periods, do NOT reset the device
        - made main line coding into a main() function and call it if
          __name__ == '__main__'
        - made proper doc-strings for functions instead of traditional
          comments before the function declaration.  (fat lot of
          good it'll do when the command name uses hyphens instead of
          underscores and doesn't end in .py)
        - added example to README.txt and README.md
        - updated man page for -m option.  added missing -x.  fixed stuff.
        - updated version in setup.py

1.3     Aug 12, 2020
        - now works with python version 3,  as well as version 2
        - is_reset_locked():
          changed returns from 0 and 1 to False and True
        - reset_device():
          variable quiet_flag was not available and caused a traceback
          from NameError. made quiet_flag a global
        - logit():
          made smarter in handling exception when open/write fails
        - num_too_big():
          new function created.
          Used to test values for number of ping tries (-t), reset time (-r),
          wait time (-w) and ping timeout (-x)
        - usage():
          complete overhaul.  much easier to read source
        - refactoring:
          line lengths now less than 80 characters.
          some info and error messages changed for shorter line lengths.
          logic changed to effect logging and messaging if force flag is set

2.0     Nov 3, 2023
        COMPLETE OVERHAUL.

        The entire package has been reorganized to utilize modern best
        practices for Python packaging and distribution.  The program has
        now been changed to be a module and uses the main() function as
        an entry point for a wrapper program that imports and calls it.
        It is no longer using setup() and disttools, but instead has pip
        as the front-end for a build, and uses a back-end of hatchling,
        as defined in the pyproject.toml file.  

        - added option -L/--lockfile option
        - added --dns-hosts to be accepted along with -H and --hosts
          for better clarity
        - added default dns host of 8.8.4.4 to go along with 8.8.8.8
        - functions moved to a separate module.  A lot of refactoring.

2.0.1   Nov 14, 2023
        Makefile:
            - removed the --editable from install
        src/pi_power_relay_moxad/__init__.py,pyproject.toml:
            - updated version
        README.txt,README.md:
            - changed example from using --hosts to --dns-hosts

2.0.2   Nov 16, 2023
       - changed the --dns-hosts back to --hosts in the README files.
         While the default hosts pinged happen to be DNS hosts, they
         don't have to be.  Just any device that responds to pings.
       - changed program name from full pathname to just filename.
         So now pi-power-relay instead of /a/b/c/pi-power-relay
       - fixed the year in the date of the above 2.0.1 RELEASE-NOTES entry
         from 2013 to 2023
       - updated man-page to show correct defaults of host(s) pinged


src/pi_power_relay_moxad/functions.py
import os
import sys
import time
import re

from . import globals

# see if we are running on a raspberry Pi
try:
    if os.uname()[4].lower().startswith( 'arm' ):
        import RPi.GPIO as GPIO
except ModuleNotFoundError as err:
    sys.stderr.write( "Error: missing module: %s\n" % err )
    sys.exit(1)


def num_too_big( num, max ):
    """Test if a number is too large

    Arguments:
        number being tested
        maximum value allowed
    Returns:
        False:  ok
        True:   number is too large
    """

    if num > max:
        return( True )
    return( False )


def dprint( str ):
    """
    print debug statement if globals.debug_flag is True

    Arguments:
        string to print
    Returns:
        0
    Exceptions:
        none
    Globals:
        globals.debug_flag
    """

    if globals.debug_flag == False:
        return None

    print( "debug: {}".format ( str ))
    return None


def ping( host, tries=3, timeout=2 ):
    """ping a host

    Arguments:
        1:  host
        2:  number of tries
        3:  timeout for ping
    Returns:
        0:  down
        1:  up
    """

    my_name = sys._getframe().f_code.co_name

    for i in range( tries ):
        response = os.system( "ping -c 1 -w" + str(timeout) + " " + 
                              host + " > /dev/null 2>&1" )

        state = "down"
        if response == 0:
            state = "up"

        msg = "ping response (timeout={0:d}) for {1:s} is {2:d} ({3:s})". \
            format( timeout, host, response, state )

        dprint( "{0:s}(): try #{1:d} {2:s}".format( my_name, i+1, msg ))

        if response == 0:
            return(1)   # up

    return(0)   # down


def test_network( hosts, tries=3, timeout=2 ):
    """test if network reachable given an array of hosts to ping

    Arguments:
        1:  array of hosts
        2:  number of pings to try for each host in host-list
        3:  timeout for ping
    Returns:
        0:  down
        1:  up
    """

    my_name = sys._getframe().f_code.co_name

    result = 0          # assume not reachable

    for host in hosts:
        dprint( "{0:s}(): testing host \'{1:s}\' with ping". \
            format( my_name, host ))
        response = ping( host, tries, timeout )
        if response == 1:
            result = 1       # host on network is reachable
            break

    state  = 'down'
    if ( result == 1 ):
        state = 'up'

    dprint( "{0:s}(): final result = {1:d} ({2:s})". \
        format( my_name, result, state ))

    return( result )


def reset_device( pin_num, wait_time, lock_file, vals={}, GPIO_active=True ):
    """reset a device

    Arguments:
        1:  GPIO pin number
        2:  wait time between pin toggles
        3:  lock-file
        4:  optional values dictionary containing:
                quiet-flag  (default = False)
                device-name (default = 'device')
        5:  flag if GPIO active.  default = True
    Returns:
        0
    """

    my_name = sys._getframe().f_code.co_name
    quiet_flag  = vals.get( 'quiet-flag', False )
    device_name = vals.get( 'device-name', 'device' )

    if GPIO_active:
        GPIO.setmode( GPIO.BCM )
        GPIO.setup( pin_num, GPIO.OUT )
    else:
        if not quiet_flag:
            print( "GPIO not active.  simulating RESET of '{0:s}'". \
                format( device_name ))

    dprint( "{0:s}(): setting PIN {1:d} HIGH". \
            format( my_name, pin_num ))
    if GPIO_active:
        GPIO.output( pin_num, GPIO.HIGH )

    dprint( "{0:s}(): sleeping for {1:d} seconds". \
        format( my_name, wait_time ))
    time.sleep( wait_time )

    dprint( "{0:s}(): setting PIN {1:d} LOW". \
            format( my_name, pin_num ))
    if GPIO_active:
        GPIO.output( pin_num, GPIO.LOW )

        # shut it all down
        GPIO.cleanup()

    write_timestamp( lock_file, vals )

    return(0)


def write_timestamp( file, vals={} ):
    """write a timestamp to a file

    time-stamp file format:
        line1:  seconds since epoch - to be read by this program
        line2:  human readable string of reset time
    Arguments:
        1:  filename
        2:  optional values dictionary containing:
                device-name (default = 'device')
    Returns:
        0
    """

    my_name = sys._getframe().f_code.co_name
    device_name = vals.get( 'device-name', 'device' )

    dprint( "{0:s}(): writing timestamp lock to {1:s}". \
            format( my_name, file ))

    now_num = int( time.time())
    now_str = time.strftime( "%a %b %d, %Y %H:%M:%S" )
    try:
        f = open( file, "w" )
        f.write( str( now_num ) + "\n" )
        f.write( "reset {} at {}\n". \
            format( device_name, now_str ))
        f.close()
    except (IOError) as err:
        pass

    return(0)


def logit( file, message ):
    """log a message

    Arguments:
        1:  filename
        2:  message
    Returns:
        1:  an error occurred
        0:  success
    """

    my_name = sys._getframe().f_code.co_name

    try:
        f = open( file, "a" )
        now = time.strftime( "%a %b %d, %Y @ %H:%M" )
        f.write( now + ": " + message )
        f.close()
    except (IOError) as err:
        err_code, err_msg  = err.args
        raise Exception( "%s(): %s: %s" % \
            ( my_name, err_msg, file ))

    return(0)


def is_reset_locked( file, wait_time ):
    """test if there is a lock in place for resetting

    reads a file previously written on last reset.
    1st line is Unix epoch timestamp.

    Arguments:
        1:  lock filename
        2:  number seconds to honour lock since last reset
    Returns:
        False:  no lock
        True:   locked
    """

    my_name = sys._getframe().f_code.co_name
    try:
        f = open( file, "r" )
        dprint( "{}(): checking timestamp in {}".format( my_name, file ))

        # get the last reset timestamp - its on the 1st line
        last_time = int( f.readline().rstrip())
        if is_int( last_time ) == False:
            return( False )   # give up - no lock

        now = int( time.time())
        diff = now - last_time
        f.close()

        dprint( "{0:s}(): Got last reset timestamp of {1:d}". \
            format( my_name, last_time ))
        dprint( "{0:s}(): {1:d} seconds since last reboot". \
            format( my_name, diff ))

        if diff < wait_time:
            dprint( "{0:s}(): reset lock in place. {1:d} < {2:d} seconds". \
                format( my_name, diff, wait_time ))
            return( True )   # locked

    except (IOError) as err:
        pass
        return( False )       # no lock

    return( False )


def is_int(s):
    """Test to see if a string is an integer

    Arguments:
        1:  string
    Returns:
        True:   an integer
        False:  not an integer
    """

    try:
        int(s)
        return True
    except ValueError:
        return False


def convert_times( time_range ):
    """convert a string of HH:MM-HH:MM and return ( start-time, end_time )

    the return times are HH * 60 + MM for the number of seconds into the 
    current day.  You can't cross over a 24 hour boundary.
    ie: you can't say 22:30-01:30. 
    Hours are given in 24-hour format.
    There is no distinction for AM and PM.

    Arguments:
        1:  HH:MM-HH:MM
    Returns:
        ( start-time-mins, end-time-mins )
    Exceptions:
        Exception
    """

    range = re.compile( '^(\d+:\d+)-(\d+:\d+)$' )
    m = range.match( time_range )
    if m:
        try:
            start_time = m.group(1).split(':')
            end_time   = m.group(2).split(':')

            s_hr  = int( start_time[0] )
            s_min = int( start_time[1] )
            e_hr  = int( end_time[0] )
            e_min = int( end_time[1] )
        except TypeError as err:
            raise Exception( err )

        if ( e_hr < s_hr ):
            err = "maintenance time range given has end hour before start hour"
            raise Exception( err )

        if ( e_hr == s_hr ) and ( e_min < s_min ):
            err = "maintenance time range given has end time before start time"
            raise Exception( err )

        if ( s_hr > 24 ) or ( e_hr > 24 ) or ( s_min > 59 ) or ( e_min > 59 ):
            err = "nonsense maintenance time given"
            raise Exception( err )

        maint_start = s_hr * 60 + s_min
        maint_stop  = e_hr * 60 + e_min
        return( maint_start, maint_stop )
    else:
        err = "maintenance time range not of format: HH:MM-HH:MM"
        raise Exception( err )


src/pi_power_relay_moxad/pi_power_relay.py
"""pi-power-relay

see if host(s) on an outside network are reachable via ping.
If not, reset a device, likely a modem, that is known to be flakey,
by power cycling it via a power relay.

This is intended to be run on a Raspberry Pi, using an 'IoT Power Relay
version 2' from Digital Loggers, to use GPIO to power-cycle a device.
The power relay is plugged into a UPS, while the Raspberry Pi is plugged
into the 'always ON' power port and the modem is plugged into the
'normally ON' power port.  The program, when detecting a loss of network
connectivity, will drop power on the 'normally ON' port, and bring it 
back after a short period.  See the product it uses at:
  https://dlidirect.com/products/iot-power-relay
Typically run from the cron every couple of minutes.

If run on something other than a Raspberry Pi, it will assume it is
a development or test box and will NOT do the actual shutdown, but
will still do everything else.  On the Raspberry Pi, you will need the
RPi.GPIO module installed, by either doing a:
    sudo apt-get install rpi.gpio
on a Debian based Linux box, or installing from the Python Package Index:
        pip install RPi.GPIO
This module will only be imported if run on a Raspberry Pi.

There is an option called -e or --delay-exit, which may not be clear why.
The reason for this is that if you do not use the --quiet option, then
an informational message is printed out before the program exits.  But if
you are relying on ouput from the crontab to then be sent to you on some
other machine via e-mail, it will fail if you do not have a full blown
mailer set up that queues the message, and the DNS lookup for the
destination will fail because the device (modem?) has not
recovered yet.  So, this delay will give the device time to recover,
before you spit out the message and then exit.
The default is no delay.  A good value to use would be 180 seconds.
"""

import os
import sys
import time
import re
import getpass

from . import __version__
from .globals import progname, debug_flag
from .functions import *


def die( error ):
    """print an error message and exit

    Arguments:
        1:  message
    Globals:
        globals.progname
    """

    if globals.progname == None:
        prefix = ''
    else:
        prefix = globals.progname + ': '

    sys.stderr.write( "{}{}\n".format( prefix, error ))
    sys.exit(1)


# main
#
# Arguments:
#   command-line arguments
# Returns:
#   0:  ok
#   1:  not ok
# Exceptions:
#   none

def main( argv=sys.argv ):
    """main program

    use -h or --help for command line options

    Returns:
        0:  ok
        1:  not ok
    """

    progname = os.path.basename( argv[0] )
    if progname == None or progname == "":
        progname = 'pi_power_relay'

    globals.progname = progname     # make available to other functions

    # see if we are running on a raspberry Pi
    HAVE_GPIO = False
    if os.uname()[4].lower().startswith( 'arm' ):
        HAVE_GPIO = True

    # values that can be changed via options

    pin_number       = 25            # GPIO pin number
    reset_time       = 15            # between setting pin HIGH, then LOW
    wait_time        = 10 * 60       # wait time to reset again 
    ping_timeout     = 2             # secs to wait for ping to time out
    ping_tries       = 3             # number of pings to try
    delay_exit_wait  = 0             # delay before output message and exit
    device_name      = 'device'      # use device name in log message
    log_file         = ""            # LOG file.  none by default
    maint_times      = []            # array of maint times HH:MM-HH:MM
    quiet_flag       = False
    force_flag       = False
    help_flag        = False
    logging_flag     = False
    dns_hosts        = [ '8.8.4.4', '8.8.8.8' ]
    lock_file        = "/tmp/pi-power-relay--reset-time"

    # limits

    max_reset_time   = 60
    max_wait_time    = 30 * 60
    max_ping_timeout = 10
    max_ping_tries   = 10

    # get options

    num_args = len( argv )
    i = 1
    while i < num_args:
        try:
            arg = argv[i]
            if arg == '-d' or arg == '--debug':
                globals.debug_flag = True
            elif arg == '-h' or arg == '--help':
                help_flag = True
            elif arg == '-f' or arg == '--force-reset':
                force_flag = True
            elif arg == '-q' or arg == '--quiet':
                quiet_flag = True
            elif arg == '-r' or arg == '--reset-time':
                i = i + 1 ; val = argv[i]
                if is_int( val ) == False:
                    die( "Not an integer: \'{0:s}\'".format( val ))
                if ( num_too_big( int( val ), max_reset_time )):
                    die( "reset time too large ({:s} > {:d})". \
                        format( val , max_reset_time ))
                reset_time = int( val )
            elif arg == '-w' or arg == '--wait-time':
                i = i + 1 ; val = argv[i]
                if is_int( val ) == False:
                    die( "Not an integer: \'{:s}\'".format( val ))
                if ( num_too_big( int( val ), max_wait_time )):
                    die( "wait time too large ({:s} > {:d})". \
                        format( val, max_wait_time ))
                wait_time = int( val )
            elif arg == '-m' or arg == '--maint':
                i = i + 1
                maint_times.append( argv[i] )
            elif arg == '-t' or arg == '--tries':
                i = i + 1 ; val = argv[i]
                if is_int( val ) == False:
                    die( "Not an integer: \'{0:s}\'".format( val ))
                if ( num_too_big( int( val ), max_ping_tries )):
                    die( "num ping tries too large ({:s} > {:d})". \
                        format( val, max_ping_tries ))
                ping_tries = int( val )
            elif arg == '-e' or arg == '--delay-exit':
                i = i + 1 ; val = argv[i]
                if is_int( val ) == False:
                    die( "Not an integer: \'{0:s}\'".format( val ))
                delay_exit_wait = int( val )
            elif arg == '-x' or arg == '--ping-timeout':
                i = i + 1 ; val = argv[i]
                if is_int( val ) == False:
                    die( "Not an integer: \'{0:s}\'".format( val ))
                if ( num_too_big( int( val ), max_ping_timeout )):
                    die( "ping timeout too large ({:s} > {:d})". \
                        format( val, max_ping_timeout ))
                ping_timeout = int( val )
            elif arg == '-p' or arg == '--pin':
                i = i + 1 ; val = argv[i]
                if is_int( val ) == False:
                    die( "Not an integer: \'{0:s}\'".format( val ))
                val = int( val )
                if ( val > 27 ) or ( val < 0 ):
                    die( "invalid pin num: \'{}\'".format( val ))
                pin_number = val
            elif arg == '-D' or arg == '--device-name':
                i = i + 1 ; device_name = argv[i]
            elif arg == '-L' or arg == '--lockfile':
                i = i + 1 ; lock_file = argv[i]
            elif arg == '-l' or arg == '--logfile':
                i = i + 1 ; log_file = argv[i]
                logging_flag = True
            elif arg == '-H' or arg == '--hosts' or arg == '--dns-hosts':
                # permit --dns-hosts and --hosts for backward compatibility
                i = i + 1 ; val = argv[i]
                dns_hosts = val.split( "," )
            elif arg == '-V' or arg == '--version':
                print( "version: {0}".format( __version__ ))
                return(0)
            else:
                m = re.match( "^\-", arg )
                if m:
                    die( "unknown option: \'%s\'" % arg )
                else:
                    die( "unknown option: \'%s\'" % arg )

        except IndexError as err:
            msg = "Missing argument value to \'{0:s}\'?".format( arg )
            die( "%s.  %s" %  ( str(err), msg ))

        i = i+1

    # now that we have processed all our options, our usage can show
    # defaults properly

    if help_flag:
        print( "usage: {} [options]*".format( progname ))

        options = """\
        [-d|--debug]               debugging output
        [-e|--delay-exit num]      delay time before output and exit ({} secs)
        [-f|--force-reset]         reset now and quit, despite state or lock
        [-h|--help]                print this help info
        [-l|--logfile string]      log filename (none by default)
        [-m|--maint HH:MM-HH:MM]*  Maintenance time to NOT reset (none by default)
        [-p|--pin num ]            GPIO pin number ({})
        [-q|--quiet]               supress informational messages
        [-r|--reset-time num]      wait time between GPIO state change ({} secs)
        [-t|--tries num]           max number of ping attempts per host ({})
        [-w|--wait-time num]       reset wait time after previous reset ({} secs)
        [-x|--ping-timeout num]    wait time for ping to time out ({} secs)
        [-D|--device-name string]  name of thing being reset for log ({})
        [-H|--hosts string(s)]     comma-delimited hosts to ping ({})
        [-L|--lockfile string]     lock-file ({})
        [-V|--version]             print version of this program ({})\
        """
        print( options.format( delay_exit_wait, pin_number, reset_time,
            ping_tries, wait_time, ping_timeout, device_name,
            ','.join( dns_hosts ), lock_file, __version__ ))

        return(0)

    # if we were given any maintenance time intervals
    if len( maint_times ):
        current_time = time.localtime( None )
        c_total_mins = current_time[3] * 60 + current_time[4]
        dprint( "Current number of minutes into today is {0:d}". \
            format( c_total_mins ))

        for maint_str in maint_times:
            try:
                ( maint_start, maint_end ) = convert_times( maint_str )
            except Exception as err:
                die( err )
            dprint( "maint start={0:d}, end={1:d} (mins) for period {2:s}". \
                format( maint_start, maint_end, maint_str ))

            if ( c_total_mins > maint_start ) and ( c_total_mins < maint_end ):
                dprint( "now in maintenance interval of {0:s}.  Quitting.". \
                    format( maint_str ))
                return(0)
            else:
                dprint( "not in maintenance interval of %s." % maint_str )

    # See if we are running on our target Raspberry pi.
    # Anything else will be a development or test environment

    if HAVE_GPIO == True:
        dprint( "Running on a Raspberry Pi." )

        # need to be root for access to memory
        if getpass.getuser() != 'root':
            die( "Need to be root to use GPIO" )
    else:
        # development/debugging code
        dprint( "I'm NOT running on a Raspberry Pi (" + os.uname()[4] + ")"  )
        dprint( "Assuming a development/test box - will NOT shutdown device" )

    # see if the network still reachable

    state = test_network( dns_hosts, ping_tries, ping_timeout )
    if state:
        # still up
        dprint( "Nothing to do. All is well.  exiting." )
        return(0)

    # see if the device is locked from a recent reset
    device_locked = False
    try:
        device_locked = is_reset_locked( lock_file, wait_time )
    except Exception as err:
        pass

    if device_locked:
        dprint( "found a timing lock: %s" % lock_file )
        # the device is locked from resetting
        if force_flag == False:
            dprint( "timing lock in effect.  skipping the reset" )
            return(0)
        else:
            # force the reset despite the lock
            dprint( "over-riding device timing lock because of force flag" )
    else:
        dprint( "no device timing lock found" )

    # ok, let's do it...
    if ( logging_flag ):
        msg = "{0:s}: network unreachable.  resetting {1:s}\n". \
            format( progname, device_name )
        try:
            logit( log_file, msg )
        except Exception as err:
            sys.stderr.write( "%s: %s\n" % ( progname, err ))
            # keep going

    # been long enough since last reset.  do it now
    try:
        vals = { 'quiet-flag': quiet_flag, 'device-name': device_name }
        reset_device( pin_number, reset_time, lock_file, vals, HAVE_GPIO )
    except Exception as err:
        sys.stderr.write( "%s: error resetting device: %s\n" % \
            ( progname, err ))
        # keep going

    # Informational message to print
    # Get the real time first before a potential delay exit

    msg = "{0:s} reset at {1:s}". \
        format( device_name, time.strftime( "%a %b %d, %Y %H:%M:%S" ))

    if ( delay_exit_wait ):
        dprint( "delay exit.  waiting {0:d} seconds".format( delay_exit_wait ))
        time.sleep( delay_exit_wait )

    if not quiet_flag:
        print( msg )

    return(0)
